src/preprocess/dataset_split/audio_data_balancer.py

```diff
@@ -0,0 +1,174 @@
+                                                                         orig_filename_waveform_sample_rate[file][1], 
+                                                                         processed_wav]
+                
+                missing_samples -= 1
+
+        for filename in orig_filename_waveform_sample_rate:
+            sf.write(filename + '.ogg', orig_filename_waveform_sample_rate[filename][2], 
+                     orig_filename_waveform_sample_rate[filename][1], format='ogg', subtype='vorbis')
+
+        for filename in augmented_filename_waveform_sample_rate: 
+            sf.write(filename + '.ogg', augmented_filename_waveform_sample_rate[filename][2], 
+                     augmented_filename_waveform_sample_rate[filename][1], format='ogg', subtype='vorbis')
+
+
+    def balance_dataset(self, direct_class_dist: dict=None):
+        """
+        :param direct_class_dist: dict, the dictionary holding the number of samples for each class.
+
+        For each bird class do the balancing by either use the waveforms of oversampled classes directly 
+        or oversample undersampled classes
+        """
+        if direct_class_dist:
+            self.class_dist = direct_class_dist
+        self._create_dirs_for_balanced_data()
+        
+        for bird_class in self.class_dist:
+            print(bird_class)
+            if self.class_dist[bird_class] >= self.target_sample_size:
+                self._preprocess_from_oversampled_classes(bird_class)
+            else:
+                self._preprocess_for_undersampled_classes(bird_class)
+
+
+
+    def get_class_dist_from_metadata(self, metadata_filepath: str, class_column: str):
+        """
+        :param metadata_filepath: str, filepath to the metadata file.
+        :param class_column: str, specifies the column where the classes are stored in.
+
+        Extract the class distribution from a metadata csv-file.
+        """
+        intermediate_class_dist = {}
+        # currently only csv metadata files supported
+        metadata_file = pd.read_csv(metadata_filepath)
+        df = metadata_file[class_column].value_counts().to_frame().reset_index().rename(columns={'index':'class_name', 'values':'count'})
+        df_dict_list = df.to_dict(orient='records')
+        for key_value_pair in df_dict_list:
+            intermediate_class_dist[key_value_pair['class_name']] = key_value_pair[class_column]
+        
+        self.class_dist = intermediate_class_dist
+
+
+if __name__ == "__main__":
+    # example on hwo to use it
+    DATASET_BASE_FILE_PATH = r"D:\Datasets\birdclef-2023"
+    TRAIN_SET_FILE_DIR = r"\train_audio"
+    TARGET_DIR = r"\train_audio_balanced"
+    metadata_file = os.path.join(DATASET_BASE_FILE_PATH + r"\train_metadata.csv")
+
+    audio_balancer = AudioDataBalancer(target_sample_size=10,
+                                       target_wav_duration=10,
```

Comment:
Why had we agreed on 10 again if the model is supposed to predicat 5 second pieces? Would only make sense if you stack several inputs or not?

Reply:
It is just an example value, can be anything